yarn-project/archiver/src/archiver/archiver.ts

```diff
@@ -136,10 +137,22 @@ export class Archiver implements L2BlockSource, UnverifiedDataSource, ContractDa
       currentBlockNumber,
       this.lastProcessedBlockNumber + 1n, // + 1 to prevent re including messages from the last processed block
     );
+    const retrievedCancelledL1ToL2Messages = await retrieveNewCancelledL1ToL2Messages(
+      this.publicClient,
+      this.inboxAddress,
+      blockUntilSynced,
+      currentBlockNumber,
+      this.nextL2BlockFromBlock + 1n,
```

Comment:
Related, anyone know the reasoning behind having `nextL2BlockFromBlock`? Seems like it is updated based on the retrievedBlocks but not clear to me why it would need to differ from `lastProcessedBlockNumber`? When based off the blocknumber when the latest l2block was published will search over the same logs again for no good reason? e.g., if 400 blocks were appended after we are still storing the n-400 block as where we got to but we know there was nothing in those 400 blocks.

Reply:
I added `lastProcessedBlockNumber` in my last pr to make sure that we scan messages from the most recently scanned block rather than the last block with a rollup as it caused them to be indexed twice. You're right though, `nextL2BlockFromBlock` can likely be removed now, and all replaced with `lastProcessedBlockNumber`

Reply:
ah not sure how I missed that rip

Reply:
The reason we use `nextL2BlockFromBlock` is that eventually consistent load balanced providers like Infura will give you false negatives. So they may correctly tell you that the current ETH block number is 1000 but that there are no `Rollup` logs in that block when in fact there are, it's just the node you happened to be routed to didn't have them yet. If you take this negative as being true and only then scan from block 1001 you will never receive the event. By repeatedly scanning events from the ETH block at which we last definitely received a rollup, we leave nothing to chance.

I'm not sure how you are publishing the L1/L2 message events. With the unverified data, we publish it first, before the block and we include in the event the call data hash of the L2Block. This means the Archiver can determine which unverified data to accept along with the new block. This is necessary in a multi-sequencer environment where there could be multiple blocks of unverified data produced by different sequencers but only 1 L2Block. We also ignore all unverified data events where the L2Block isn't yet published, this would presumably solve your double-indexing problem @Maddiaa0? We then know that once the L2Block is published, all other events related to that block will have been published and all events for the next bock must come after it.

Reply:
- users create the l1 to l2 messages that are being indexed here, so we wont be able to include a block hash to check whether they have been picked up or not, the user won't know it at that point in time.
- To get around the eventual consistency problem, we could scan the chain x blocks behind to give the providers time to catch up, not a clean solution, but in general we should be scanning behind the head of the chain anyway to account for possible reorgs

Reply:
Instead of starting the search at the last value, could probably use the last finalised or last safe block as the starting point in case last > safe? Would assume that using this would give infure sufficient time.